medcouple n log n

hmustafamail opened the pull request:
I am following up on issue #9570 ([link to issue](https://github.com/statsmodels/statsmodels/issues/9570)) by creating this pull request, which implements Medcouple in O(N log N) time. 

Legacy functionality with O(N**2) time is preserved with the use of a flag ``use_fast``. This flag defaults to the new behavior.

Please let me know if you have any questions, or if you need anything else. 

Overview:
- added fast algorithm
- added legacy multiplexing option
- revised tests
- suggested draft release note

- [ ] closes #9570 
- [ ] tests added / passed. 
- [ ] code/documentation is well formatted.  
- [ ] properly formatted commit message. See 
      [NumPy's guide](https://docs.scipy.org/doc/numpy-1.15.1/dev/gitwash/development_workflow.html#writing-the-commit-message).

bashtage commented on statsmodels/stats/stattools.py:
```diff
@@ -441,7 +444,220 @@ def _medcouple_1d(y):
+        mid = (beg + end) // 2
+        trial = AW[mid][0]
+
+        wleft = sum(w for a, w in AW if a < trial)
```
These might also be slow since using Python objects.

hmustafamail commented:
Thank you for your code critique. I made the following main changes:

- Replaced Python lists with NumPy arrays
- Removed zip()
- Added NaN check to the top of _medcouple_nlogn()

hmustafamail commented:
AI was used to help translate the contribution from Python 2 to Python 3 and to revise it for project standards. I personally reviewed, comprehended, and validated all changes by hand.

bashtage commented:
After many months, a big thanks for this.

hmustafamail commented:
@bashtage Thank you for this wonderful news as well as for your guidance and service to our community.
